rfcs/0049-flakes.md

```diff
@@ -0,0 +1,717 @@
+
+    inputs.grcov = {
+      url = github:mozilla/grcov;
+      flake = false;
+    };
+
+    outputs = { self, nixpkgs, grcov }: {
+      packages.grcov = stdenv.mkDerivation {
+        src = grcov;
+        ...
+      };
+    };
+
+## Flake references
+
+Flake references are a vaguely URL-like syntax to specify which flake
+to use. This is used on the command line (e.g. in `nix build
+nixpkgs:hello`, `nixpkgs` is a flake reference), and in the list of
+flake dependencies in `flake.nix` (e.g. in `inputs.nixpkgs = {};`).
+
+Currently the following types of flake references are supported:
+
+* Git repositories. These have the form
+
+      git(+http|+https|+ssh|+git|+file|):(//<server>)?<path>(\?<params>)?
+
+  `<params>` are a list of key/value pairs in URI query parameter
+  syntax. The following parameters are supported:
+
+  * `ref`: The branch or tag to fetch. The default is `master`.
+  * `rev`: The Git commit hash to fetch. Note that this commit must be
+    an ancestor of `ref`, since Nix doesn't clone the entire
+    repository, only the specified `ref` (and the Git protocol doesn't
+    allow fetching a `rev` without a known `ref`). The default is the
+    commit currently pointed to by `ref`.
+  * `dir`: The subdirectory of the repository in which `flake.nix` is
+    located. This parameter enables having multiple flakes in a
+    repository. The default is the root directory.
+
+  For example, the following are valid Git flake references:
+
+  * `git+https://example.org/my/repo`
+  * `git+https://example.org/my/repo?dir=flake1`
+  * `git+ssh://git@github.com:NixOS/nix?ref=v1.2.3`
+  * `git://github.com/edolstra/dwarffs?ref=unstable&rev=e486d8d40e626a20e06d792db8cc5ac5aba9a5b4`
+  * `git+file:///home/my-user/some-repo/some-repo`
+
+* Local paths. These have the form
+
+      <path>(\?<params)?
+
+  where `<path>` must refer to (a subdirectory of) a Git repository.
+  These differ from `file://` flake references in a few ways:
+
+  * They refer to the working tree (unless an explicit `rev` or `ref`
+    is specified), so evaluation can access dirty files. (Dirty files
+    are files that are tracked by git but have uncommitted changes.)
+
+  * The `dir` parameter is automatically derived. For example, if
+    `/foo/bar` is a Git repository, then the flake reference
+    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.
+
+* GitHub repositories. These are downloaded as tarball archives,
+  rather than through Git. This is often much faster and uses less
+  disk space since it doesn't require fetching the entire history of
+  the repository. On the other hand, it doesn't allow incremental
+  fetching (but full downloads are often faster than incremental
+  fetches!). The syntax is:
+
+      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?
+
+  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
+  commit hash (`rev`). Note that unlike Git, GitHub allows fetching by
+  commit hash without specifying a branch or tag.
```

Comment:
According to https://github.com/NixOS/nixos-channel-scripts/pull/27#issuecomment-547376718 flakes can’t handle slashes in branch names.

I think flakes not working with arbitrary branch names will be extremely annoying, and would be a mistake.

Reply:
Strictly speaking you can use slashes, but you have to write something like `?ref=refs/heads/channels/nixos-19.09`.

Reply:
I suppose that’s fine, as long as it’s made extremely obvious to the user what they’ve done wrong if they try to use a slash and it doesn’t work.

Reply:
Would also be useful to have `nix flake simplify-url` or something so that if you were generating a flake URI, you could just always use `ref`, full URLs, etc, and have a way of converting it to a simpler form if possible, rather than having to account for all these weird bits yourself.

But even then it’s starting to feel like quite a big ask for a human or computer to be able to construct these.

Reply:
Looking at the RFC, it doesn't really seem clear *why* slashes in a `github:` uri aren't allowed? For local paths, it says the `dir` parameter is automatically derived; it doesn't read like this would be the case for git{,hub} uris.